internal/contentlog/contentlog_json_writer.go

```diff
@@ -0,0 +1,361 @@
+// Package contentlog provides a JSON writer that can write JSON to a buffer
+// without any memory allocations and Logger that can write log entries
+// with strongly-typed parameters.
+package contentlog
+
+import (
+	"strconv"
+	"time"
+
+	"github.com/kopia/kopia/internal/freepool"
+)
+
+var commaSeparator = []byte(",")
+
+const (
+	decimal     = 10
+	hexadecimal = 16
+)
+
+// JSONWriter is a writer that can write JSON to a buffer
+// without any memory allocations.
+type JSONWriter struct {
+	buf       []byte
+	separator []byte
+
+	separatorStack [][]byte
+}
+
+// ParamWriter must be implemented by all types that write a parameter ("key":value)to the JSON writer.
+type ParamWriter interface {
+	WriteValueTo(sw *JSONWriter)
+}
+
+func (jw *JSONWriter) beforeField(key string) {
+	jw.buf = append(jw.buf, jw.separator...)
+	jw.buf = append(jw.buf, '"')
+	jw.buf = append(jw.buf, key...)
+	jw.buf = append(jw.buf, '"', ':')
+	jw.separator = commaSeparator
+}
+
+func (jw *JSONWriter) beforeElement() {
+	jw.buf = append(jw.buf, jw.separator...)
+	jw.separator = commaSeparator
+}
+
+// object
+
+// BeginObjectField starts an object field.
+func (jw *JSONWriter) BeginObjectField(key string) {
+	jw.beforeField(key)
+	jw.BeginObject()
+}
+
+// BeginObject starts an object.
+func (jw *JSONWriter) BeginObject() {
+	jw.buf = append(jw.buf, '{')
+	jw.separatorStack = append(jw.separatorStack, commaSeparator)
+	jw.separator = nil
+}
+
+// EndObject ends an object.
+func (jw *JSONWriter) EndObject() {
+	jw.separator, jw.separatorStack = jw.separatorStack[len(jw.separatorStack)-1], jw.separatorStack[:len(jw.separatorStack)-1]
```

Comment:
split these into 2 lines to make it more readable, no?

Reply:
done